url71665d7d68_branscha_lib_scripty_tgz-pJ8-com_sdicons_scripty_ScriptyAutoReplJ8/src/com/sdicons/scripty/cmdlib/BeanLibrary.java

```diff
@@ -0,0 +1,560 @@
+            if("/".equals(currentDirectory)) lAbsPath = currentDirectory + aPath;
+            else lAbsPath = currentDirectory + "/" + aPath;
+            return resolve(lAbsPath, aCtx);
+        }
+    }
+
+    private static Object resolveCanonical(List<String> aPieces, int aCurrPiece, Object aBase)
+    throws CommandException
+    {
+        if(aPieces.size() <= aCurrPiece) return aBase;
+        final String lPiece = aPieces.get(aCurrPiece);
+        int lIdx = -1;
+        try
+        {
+            lIdx = Integer.parseInt(lPiece);
+        }
+        catch (NumberFormatException e1)
+        {
+            // Ignore this exception!
+        }
+
+        if(aBase instanceof IContext && lIdx < 0)
+        {
+            final IContext lCtx = (IContext) aBase;
+            if(lCtx.isBound(lPiece))
+                return resolveCanonical(aPieces, aCurrPiece + 1, lCtx.getBinding(lPiece));
+            else
+                throw new CommandException(String.format("Cannot find '%s' in context.", lPiece));
+        }
+        else if(aBase instanceof Map && lIdx < 0)
+        {
+            final Map lMap = (Map) aBase;
+            if(lMap.containsKey(lPiece))
+                return resolveCanonical(aPieces, aCurrPiece + 1, lMap.get(lPiece));
+            else
+                throw new CommandException("ERROR - map.");
+        }
+        else if(aBase instanceof List)
+        {
+            final List lList = (List) aBase;
+            if(lIdx < 0) throw new CommandException("List expects an index.");
+            if(lIdx >= lList.size()) throw new CommandException("List index out of range.");
+            else return resolveCanonical(aPieces, aCurrPiece + 1, lList.get(lIdx));
+        }
+        else  if(aBase != null)
+        {
+            if(aBase.getClass().isArray())
+            {
+                // Array rendering.
+                ///////////////////
+
+                if(lIdx >=0 && lIdx < Array.getLength(aBase))
+                    return resolveCanonical(aPieces, aCurrPiece + 1, Array.get(aBase, lIdx));
+                else throw new CommandException("ERROR - index out of range/wrong index type/no index.");
+            }
+            else
+            {
+                try
+                {
+                    // Try bean.
+                    //
+                    final BeanInfo lInfo = Introspector.getBeanInfo(aBase.getClass());
+                    final PropertyDescriptor lProps[] = lInfo.getPropertyDescriptors();
+                    for (PropertyDescriptor lProp : lProps)
+                    {
+                        if (lProp.getName().equals(lPiece))
+                        {
+                            Method lGetter = lProp.getReadMethod();
+
+                            Object[] lArgs;
+                            if(lIdx >= 0) lArgs = new Object[]{lIdx};
+                            else  lArgs = new Object[]{};
+
+                            Object lVal = lGetter.invoke(aBase, lArgs);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Catch `InvocationTargetException` explicitly when you call `Method.invoke()`. Use `Throwable.getCause()` or `getTargetException()` in the catch handler to access more information about the underlying exception. See the following diffs on GitHub that make this code change: [diff1](https://github.com/EsfingeFramework/gamification/commit/20792047e6c69457cca03e877f2ada578188f32a#diff-6808c12fe9ddaea042db84f95dd5ba78adab99d0cb18bb4d85916a8723530768L21), [diff2](https://github.com/cucumber-attic/cuke4duke/commit/f4821d400cf2590126f1298a8dff9f7a33d45dc5#diff-80adaffc20c0e6e7d2e42312386ffe1ddc10213bf5b8e8919895fd85d666c8deL34)

Learn more: [InvocationTargetException in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/InvocationTargetException.html)